[now-build-utils] Add framework detector functions

TooTallNate opened the pull request:
As described in [PRODUCT-399]. This is similar in nature to the current framework detection logic in static-build, however the detectors are now functions which allow for detection of frameworks without a `package.json` file (i.e. jekyll), and also includes detection of build and dev commands for use in project settings.

See the unit tests for an example implementation of `DetectorFilesystem` backed by a real local filesystem using the `fs-extra` module.

[PRODUCT-399]: https://zeit.atlassian.net/browse/PRODUCT-399

TooTallNate committed:
[now-build-utils] Add framework detector functions

As described in [PRODUCT-399]. This is similar in nature to the
current framework detection logic in static-build, however the detectors
are now functions which allow for detection of frameworks without
a `package.json` file (i.e. jekyll), and also includes detection of
build and dev commands for use in project settings.

TooTallNate committed:
Dedup `aggregate-error` dep in yarn.lock

styfle commented on packages/now-build-utils/.gitignore:
```diff
@@ -1,3 +1,4 @@
+?.js
```
What does this mean?

![double rainbow](https://media1.tenor.com/images/04dd032f5547f668c7d6147997bf6803/tenor.gif?itemid=11850694)

TooTallNate replied:
I have ephemeral development/testing files like `d.js`. This way they don't get committed by accident.

styfle commented on yarn.lock:
```diff
@@ -2221,13 +2221,13 @@ agentkeepalive@^3.4.1:
-  version "3.0.0"
-  resolved "https://registry.yarnpkg.com/aggregate-error/-/aggregate-error-3.0.0.tgz#5b5a3c95e9095f311c9ab16c19fb4f3527cd3f79"
-  integrity sha512-yKD9kEoJIR+2IFqhMwayIBgheLYbB3PS2OBhWae1L/ODTd/JF/30cW0bc9TqzRL3k4U41Dieu3BF4I29p8xesA==
+aggregate-error@^3.0.0, aggregate-error@3.0.1:
```
Why did this change?

styfle replied:
Oh I see package.json now 🤦‍♂

styfle commented on packages/now-build-utils/src/detectors/filesystem.ts:
```diff
@@ -0,0 +1,99 @@
+ */
+export default abstract class DetectorFilesystem {
+  protected abstract _readFile(name: string): Promise<Buffer>;
+  protected abstract _exists(name: string): Promise<boolean>;
```
Is that a normal convention to use underscores for abstract methods like this?

TooTallNate replied:
Node.js does it, for example when subclassing `stream.Readable` you have to implement the `_read()` function.

styfle commented on packages/now-build-utils/src/detectors/filesystem.ts:
```diff
@@ -0,0 +1,99 @@
+
+async function nullEnoent<T>(p: Promise<T>): Promise<T | null> {
+  try {
+    return await p;
```
Is `await` necessary here? It seems like the return type is a promise.

TooTallNate replied:
It is necessary because otherwise the `try/catch` is short-circuited (i.e. if an error is thrown, it won't be caught in this function without the `await`.

styfle commented on packages/now-build-utils/src/detectors/filesystem.ts:
```diff
@@ -0,0 +1,99 @@
+    if (!pkg) return false;
+    if (pkg.dependencies && pkg.dependencies[name]) return true;
+    if (pkg.devDependencies && pkg.devDependencies[name]) return true;
+    return false;
```
Just a nit pick here but you might want to use `hasOwnProperty` and you can probably simplify:

```ts
const pkg = await this.readPackageJson();
const { dependencies = {}, devDependencies = {} } = pkg || {};
return dependencies.hasOwnProperty(name) || devDependencies.hasOwnProperty(name);
```

styfle commented on packages/now-build-utils/src/detectors/index.ts:
```diff
@@ -0,0 +1,51 @@
+    for (const p of promises) {
+      p.then(v => {
+        if (v || --unresolved === 0) {
+          resolve(v);
```
Why does it resolve even when there are no truthy values? I would expect it to reject if none are truthy. Or maybe return `null`?

TooTallNate replied:
If it's the final promise (unresolved === 0) then it needs to resolve _something_, so it resolves to the _falsy_ value of the final promise. `null` would work too for consistency, but the point is that it's already being resolved to an falsy value.

styfle reviewed (commented):
I realize that most of the zero config work has been done in `now-build-utils`, but I think this was not a great place for it.

The `@now/build-utils` used to have a very specific task and it was injected for each builder to use.

What do you think about creating a new package for this framework detection and possibly zero config?

TooTallNate commented:
> What do you think about creating a new package for this framework detection and possibly zero config?

I had this same thought. `@now/detectors` or something would make more sense to me, and also make it easier to consume in the web browser.

TooTallNate committed:
Simplify `hasDependency()`

TooTallNate committed:
Adjust hugo detect logic

TooTallNate committed:
Add "jekyll" detector

TooTallNate committed:
Add "next" detector

TooTallNate committed:
Optimize "gatsby" detector

Check the `package.json` first

TooTallNate committed:
Add optimization to check "package.json" based detectors first

They're by far the most common and only rely on checking the single
`package.json` file. If no pkg-based detector succeeds then the non-pkg
detectors are consulted afterwards.

TooTallNate committed:
Add "vue" detector

TooTallNate committed:
Add "angular" detector

TooTallNate committed:
Add "polymer" detector

TooTallNate committed:
Add "svelte" detector

TooTallNate committed:
Add "create-react-app-ejected" detector

TooTallNate committed:
Add "gridsome" detector

TooTallNate committed:
Add "umi" detector

TooTallNate committed:
Add "sapper" detector

TooTallNate committed:
Add "saber" detector

TooTallNate committed:
Add "stencil" detector

TooTallNate committed:
Merge branch 'canary' of github.com:zeit/now into add/build-utils-detectors

TooTallNate committed:
Add CRA ejected commands

TooTallNate committed:
Add "brunch" detector

TooTallNate committed:
Add "middleman" detector

TooTallNate committed:
Add missing commands

TooTallNate committed:
`buildCommand` is required

TooTallNate committed:
Add initial unit tests

TooTallNate committed:
Fix unit tests on Circle

styfle commented:
Are you planning to make `@now/detectors`?

I even think it might be better to put this detection in `@now/static-build` because nearly all the test fixtures are the same so you could avoid duplicating all of them.

styfle commented on packages/now-build-utils/package.json:
```diff
@@ -34,10 +36,12 @@
     "fs-extra": "7.0.0",
     "glob": "7.1.3",
     "into-stream": "5.0.0",
+    "jest": "24.9.0",
```
I think `jest` is only in the root `package.json` since nearly all the other packages use it.
Can we do the same for `ts-jest` and solve this problem for all packages in the monorepo?

styfle commented on packages/now-build-utils/src/detectors/jekyll.ts:
```diff
@@ -0,0 +1,15 @@
+  }
+  return {
+    buildCommand: ['jekyll', 'build'],
+    buildDirectory: '_site',
```
You need to read the `_config.yml` file to find the output directory and assign it like

```js
buildDirectory: yamlConfig['destination']
```

https://jekyllrb.com/docs/configuration/options/

styfle commented on packages/now-build-utils/src/detectors/hugo.ts:
```diff
@@ -0,0 +1,18 @@
+  }
+  return {
+    buildCommand: ['hugo'],
+    buildDirectory: 'public',
```
You need to read the config file to detect the output directory called `publishDir`.

https://gohugo.io/getting-started/configuration/

TooTallNate committed:
Merge branch 'canary' into add/build-utils-detectors
